pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -119,6 +126,106 @@ func (configgen *ConfigGeneratorImpl) BuildClusters(proxy *model.Proxy, push *mo
 	return resources, model.XdsLogDetails{AdditionalInfo: fmt.Sprintf("cached:%v/%v", cacheStats.hits, cacheStats.hits+cacheStats.miss)}
 }
 
+// BuildDeltaClusters will generate the deltas (add and delete) for a given proxy. Currently, only service changes are reflected with deltas.
+// Otherwise, we fall back onto generating everything.
+func (configgen *ConfigGeneratorImpl) BuildDeltaClusters(proxy *model.Proxy, push *model.PushContext, updates *model.PushRequest) ([]*discovery.Resource, []*discovery.Resource, model.XdsLogDetails) {
+	clusters := make([]*cluster.Cluster, 0)
+	resources := model.Resources{}
+	removedResources := model.Resources{}
+	services := make([]*model.Service, 0)
+	deletedServices := make([]*model.Service, 0)
+	envoyFilterPatches := push.EnvoyFilters(proxy)
+	cb := NewClusterBuilder(proxy, push, configgen.Cache)
+	instances := proxy.ServiceInstances
+	// delta when only services are modified and we attempt deltas
+	// we can progressively optimize this
+	log.Infof("adiprerepa: update types: %v", updates.UpdateTypes)
+	delta := updates != nil && allConfigKeysOfType(updates.ConfigsUpdated, gvk.ServiceEntry)
+	// if we can't use delta, fall back to generate all
+	if !delta || updates == nil || len(updates.ConfigsUpdated) == 0 {
+		cl, lg := configgen.BuildClusters(proxy, push)
+		return cl, make([]*discovery.Resource, 0), lg
+	}
+	configs := model.ConfigsOfKind(updates.ConfigsUpdated, gvk.ServiceEntry)
+	for s := range configs {
+		// get service that changed
+		service := proxy.SidecarScope.Service(host.Name(s.Name))
+		// SidecarScope.Service will return nil if the proxy doesn't care about the service.
+		if service == nil {
```

Comment:
it could also be that the service deleted

Reply:
I think you donot need `UpdateTypes`, you can decide which services are deleted by proxy `WatchedResource`

Reply:
so, if the service is in `WatchedResources` but the service is `nil`, we will then mark as deleted?

Do we also need to remove from `WatchedResources` ourselves?

Reply:
We should not touch WatchedResources here I think. If we need to touch WatchedResources (which, isn't it empty for CDS? Or is that only for sotw), we can do it in delta.go?

Reply:
@howardjohn WatchedResources isnt actually empty, I checked and `WatchedResources.ResourceNames` is populated with all the cluster names. I used it to cross-reference and see which cluster name was deleted, and returned that.